File: app/src/main/java/com/example/solar_alarm/CreateAlarm/AlarmScheduler.java
package com.example.solar_alarm.CreateAlarm;

import static com.example.solar_alarm.BroadcastReceiver.AlarmBroadcastReceiver.FRIDAY;
import static com.example.solar_alarm.BroadcastReceiver.AlarmBroadcastReceiver.MONDAY;
import static com.example.solar_alarm.BroadcastReceiver.AlarmBroadcastReceiver.RECURRING;
import static com.example.solar_alarm.BroadcastReceiver.AlarmBroadcastReceiver.SATURDAY;
import static com.example.solar_alarm.BroadcastReceiver.AlarmBroadcastReceiver.SUNDAY;
import static com.example.solar_alarm.BroadcastReceiver.AlarmBroadcastReceiver.THURSDAY;
import static com.example.solar_alarm.BroadcastReceiver.AlarmBroadcastReceiver.TITLE;
import static com.example.solar_alarm.BroadcastReceiver.AlarmBroadcastReceiver.TUESDAY;
import static com.example.solar_alarm.BroadcastReceiver.AlarmBroadcastReceiver.WEDNESDAY;

import android.app.AlarmManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.widget.Toast;

import androidx.annotation.RequiresApi;

import com.example.solar_alarm.BroadcastReceiver.AlarmBroadcastReceiver;
import com.example.solar_alarm.Data.Enums.OffsetTypeEnum;
import com.example.solar_alarm.Data.Tables.SolarAlarm;
import com.example.solar_alarm.Data.Tables.SolarTime;

import java.time.ZonedDateTime;
import java.util.Calendar;

public class AlarmScheduler
{
    private SolarAlarm solarAlarm;
    private SolarTime solarTime;
    private int hours;
    private int mins;
    private boolean started;

    public AlarmScheduler(SolarAlarm solarAlarm, SolarTime solarTime, int hours, int mins)
    {
        this.solarAlarm = solarAlarm;
        this.solarTime = solarTime;
        this.hours = hours;
        this.mins = mins;
        this.started = false;
    }

    @RequiresApi(api = Build.VERSION_CODES.O)
    public void schedule(Context context) throws Exception
    {


        Intent intent = new Intent(context, AlarmBroadcastReceiver.class);
        intent.putExtra(RECURRING, solarAlarm.Recurring);
        intent.putExtra(MONDAY, solarAlarm.Monday);
        intent.putExtra(TUESDAY, solarAlarm.Tuesday);
        intent.putExtra(WEDNESDAY, solarAlarm.Wednesday);
        intent.putExtra(THURSDAY, solarAlarm.Thursday);
        intent.putExtra(FRIDAY, solarAlarm.Friday);
        intent.putExtra(SATURDAY, solarAlarm.Saturday);
        intent.putExtra(SUNDAY, solarAlarm.Sunday);
        intent.putExtra(TITLE, solarAlarm.Name);

        ZonedDateTime localZonedDateTime;

        if(true) localZonedDateTime = ZonedDateTime.now().plusMinutes(mins); else // DEBUG_STATEMENT makes alarm ring immediately
            localZonedDateTime = solarTime.GetLocalZonedDateTime(solarAlarm.SolarTimeTypeId);

        if(solarAlarm.OffsetTypeId == OffsetTypeEnum.Before)
        {
            localZonedDateTime.minusHours(hours);
            localZonedDateTime.minusMinutes(mins);
        }
        else if (solarAlarm.OffsetTypeId == OffsetTypeEnum.After)
        {
            localZonedDateTime.plusHours(hours);
            localZonedDateTime.plusMinutes(mins);
        }

        Calendar calendar = Calendar.getInstance();
        calendar.setTimeInMillis(System.currentTimeMillis());
        calendar.set(Calendar.HOUR_OF_DAY, localZonedDateTime.getHour());
        calendar.set(Calendar.MINUTE, localZonedDateTime.getMinute());
        calendar.set(Calendar.SECOND, 0);
        calendar.set(Calendar.MILLISECOND, 0);

        // if alarm time has already passed, increment day by 1
        if (calendar.getTimeInMillis() <= System.currentTimeMillis())
        {
            calendar.set(Calendar.DAY_OF_MONTH, calendar.get(Calendar.DAY_OF_MONTH) + 1);
        }

        AlarmManager  alarmManager       = (AlarmManager) context.getSystemService(Context.ALARM_SERVICE);
        PendingIntent alarmPendingIntent = PendingIntent.getBroadcast(context, solarAlarm.Id, intent, 0);

        if (solarAlarm.Recurring)
        {
            String toastText = String.format("Recurring Alarm %s scheduled for %s at %02d:%02d", solarAlarm.Name, getRecurringDaysText(), localZonedDateTime.getHour(), localZonedDateTime.getMinute(), solarAlarm.Id);
            Toast.makeText(context, toastText, Toast.LENGTH_SHORT).show();

            final long RUN_DAILY = 24 * 60 * 60 * 1000;

            try
            {
                alarmManager.setRepeating(AlarmManager.RTC_WAKEUP, calendar.getTimeInMillis(), RUN_DAILY, alarmPendingIntent);
            }
            catch (Exception exception)
            {
                exception.printStackTrace();
            }
        }
        else
        {
            String toastText = String.format("One Time Alarm %s scheduled for %s at %02d:%02d", solarAlarm.Name, DayUtil.toDay(calendar.get(Calendar.DAY_OF_WEEK)), localZonedDateTime.getHour(), localZonedDateTime.getMinute(), solarAlarm.Id);;
            Toast.makeText(context, toastText, Toast.LENGTH_SHORT).show();

            try
            {
                alarmManager.setExact(AlarmManager.RTC_WAKEUP, calendar.getTimeInMillis(), alarmPendingIntent);
            }
            catch (Exception exception)
            {
                exception.printStackTrace();
            }
        }

        this.started = true;
    }

    public String getRecurringDaysText()
    {
        if (!solarAlarm.Recurring)
        {
            return null;
        }

        String days = "";
        if (solarAlarm.Monday) {
            days += "Mo ";
        }
        if (solarAlarm.Tuesday) {
            days += "Tu ";
        }
        if (solarAlarm.Wednesday) {
            days += "We ";
        }
        if (solarAlarm.Thursday) {
            days += "Th ";
        }
        if (solarAlarm.Friday) {
            days += "Fr ";
        }
        if (solarAlarm.Saturday) {
            days += "Sa ";
        }
        if (solarAlarm.Sunday) {
            days += "Su ";
        }

        return days;
    }
}

File: app/src/main/java/com/example/solar_alarm/CreateAlarm/AlarmScheduler.kt
package com.example.solar_alarm.createAlarm

import android.app.AlarmManager
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.os.Build
import android.widget.Toast
import androidx.annotation.RequiresApi
import com.example.solar_alarm.broadcastReceiver.AlarmBroadcastReceiver
import com.example.solar_alarm.data.enums.OffsetTypeEnum
import com.example.solar_alarm.data.tables.SolarAlarm
import com.example.solar_alarm.data.tables.SolarTime
import java.time.ZonedDateTime
import java.util.Calendar

@RequiresApi(Build.VERSION_CODES.O)
class AlarmScheduler(private val solarAlarm: SolarAlarm, private val solarTime: SolarTime, private val hours: Int, private val mins: Int)
{
    private var started = false

    @RequiresApi(api = Build.VERSION_CODES.O)
    @Throws(Exception::class)
    fun getIntent(context: Context) : Intent
    {
        val intent = Intent(context, AlarmBroadcastReceiver::class.java)
        intent.putExtra("SolarAlarm", solarAlarm)
        return intent
    }

    fun getCalendarInstance(localZonedDateTime: ZonedDateTime) : Calendar
    {
        val calendar = Calendar.getInstance()

        calendar.timeInMillis          = System.currentTimeMillis()
        calendar[Calendar.HOUR_OF_DAY] = localZonedDateTime.hour
        calendar[Calendar.MINUTE]      = localZonedDateTime.minute
        calendar[Calendar.SECOND]      = 0
        calendar[Calendar.MILLISECOND] = 0

        // if alarm time has already passed, increment day by 1
        if (calendar.timeInMillis <= System.currentTimeMillis())
        {
            calendar[Calendar.DAY_OF_MONTH] = calendar[Calendar.DAY_OF_MONTH] + 1
        }

        return calendar
    }

    fun getPendingIntent(context: Context, intent : Intent) : PendingIntent
    {
        try
        {
            return PendingIntent.getBroadcast(context, solarAlarm.Id, intent, PendingIntent.FLAG_IMMUTABLE)
        }
        catch (exception: Exception)
        {
            exception.printStackTrace()
            throw exception
        }
    }

    fun getLocalZonedDateTime() : ZonedDateTime
    {
        var localZonedDateTime = if (true) ZonedDateTime.now().plusMinutes(mins.toLong() + 1) else 
            solarTime.getLocalZonedDateTime(solarAlarm.SolarTimeTypeId)

        if (solarAlarm.OffsetTypeId == OffsetTypeEnum.Before)
        {
            localZonedDateTime = localZonedDateTime.minusHours(hours.toLong()).minusMinutes(mins.toLong())
        }
        else if (solarAlarm.OffsetTypeId == OffsetTypeEnum.After)
        {
            localZonedDateTime = localZonedDateTime.plusHours(hours.toLong()).plusMinutes(mins.toLong())
        }

        return localZonedDateTime
    }

    fun schedule(context: Context)
    {
        val intent             = getIntent(context)
        val localZonedDateTime = getLocalZonedDateTime()
        val calendar           = getCalendarInstance(localZonedDateTime)
        val alarmManager       = context.getSystemService(Context.ALARM_SERVICE) as AlarmManager
        val pendingIntent      = getPendingIntent(context, intent)

        if (solarAlarm.Recurring)
        {
            val toastText = String.format("Recurring Alarm %s scheduled for %s at %02d:%02d", solarAlarm.Name, recurringDaysText, localZonedDateTime.hour, localZonedDateTime.minute)
            Toast.makeText(context, toastText, Toast.LENGTH_SHORT).show()
            val RUN_DAILY = (24 * 60 * 60 * 1000).toLong()

            try
            {
                alarmManager.setRepeating(AlarmManager.RTC_WAKEUP, calendar.timeInMillis, RUN_DAILY, pendingIntent)
            }
            catch (exception: Exception)
            {
                exception.printStackTrace()
            }
        }
        else
        {
            val toastText = String.format("One Time Alarm %s scheduled for at %02d:%02d", solarAlarm.Name, localZonedDateTime.hour, localZonedDateTime.minute)
            Toast.makeText(context, toastText, Toast.LENGTH_SHORT).show()

            try
            {
                alarmManager.setExactAndAllowWhileIdle(AlarmManager.RTC_WAKEUP, calendar.timeInMillis, pendingIntent)
            }
            catch (exception: Exception)
            {
                exception.printStackTrace()
            }
        }

        started = true
    }

    private val recurringDaysText: String
        get()
        {
            var days = ""

            if (solarAlarm.Recurring)
            {
                if (solarAlarm.Monday)    days += "Mo "
                if (solarAlarm.Tuesday)   days += "Tu "
                if (solarAlarm.Wednesday) days += "We "
                if (solarAlarm.Thursday)  days += "Th "
                if (solarAlarm.Friday)    days += "Fr "
                if (solarAlarm.Saturday)  days += "Sa "
                if (solarAlarm.Sunday)    days += "Su "
            }

            return days
        }
}
